UW 572: Add make_hgrid driver (#473)
ufs-community/uwtools: docs/sections/user_guide/yaml/components/make_hgrid.rst

.. _make_hgrid_yaml:

make_hgrid
==========

Structured YAML to run :ufs-utils:`make_hgrid<make-hgrid>` is validated by JSON Schema and requires the ``make_hgrid:`` block, described below. If ``make_hgrid`` is to be run via a batch system, the ``platform:`` block, described :ref:`here <platform_yaml>`, is also required.

Here is a prototype UW YAML ``make_hgrid:`` block, explained in detail below:

.. highlight:: yaml

> .. literalinclude:: ../../../../shared/make_hgrid.yaml

Author: This may be a good place to show a comparison of the SRW YAML parameters to the UFS_UTILS parameters

UW YAML for the ``make_hgrid:`` Block
-------------------------------------

execution:
^^^^^^^^^^

See :ref:`here <execution_yaml>` for details.

run_dir:
^^^^^^^^

> The path to the directory where ``make_hgrid`` will write its outputs.

Reviewer: Sorry if I missed it, or if it's still a work in progress, but: The schema seems to permit (and conditionally require) many more keys than just these two. I don't think we should re-document them here if there's documentation we can point to, but maybe we should explicitly say that more is possible than what's documented here, and point to the authoritative docs. If our key names differ from what's in the authoritative docs, we could perhaps give a table here matching each UW YAML key name to the name used in those docs so that the user could understand the meaning of the setting and the kinds of values expected.

Author: Yeah, thats what I was thinking in addition to adding comparisons for the existing YAMLs that we're replacing, working on it

Author: Added quite a lot more key details here, let me know what you think of the formatting

YAML keys and their descriptions are taken largely verbatim from the ``make_hgrid`` tool source code, available `here <https://github.com/NOAA-GFDL/FRE-NCtools/blob/65c34e76cfa2ff567f70da83f44cd6f991315f08/tools/make_hgrid/make_hgrid.c#L71>`_.

Here is a prototype UW YAML ``make_hgrid:`` block, explained in detail below:

.. highlight:: yaml
.. literalinclude:: ../../../../shared/make_hgrid.yaml

UW YAML for the ``make_hgrid:`` Block
-------------------------------------

execution:
^^^^^^^^^^

See :ref:`here <execution_yaml>` for details.

run_dir:
^^^^^^^^

The path to the directory where ``make_hgrid`` will write its outputs.

config:
^^^^^^^

Describes the required parameters to run a ``make_hgrid`` configuration. 

grid_type:
""""""""""

Specify grid type with one of the following recognized key words:

.. list-table:: Grid Types and Requirements
   :widths: 20 80
   :header-rows: 1

   * - Grid Type
     - Requirements

>    * - from_file
>      - --my_grid_file must be specified.

Reviewer: ```suggestion
   * - ``from_file``
     - ``my_grid_file:`` must be specified.
```
I think the `--` prefixes here are confusing, as they refer to CLI switches that UW users are not directly exposed to. I think the suggested formatting would help the user connect this advice to the `my_grid_file:` section in the docs below.

Same suggestion applies to all the `--` CLI switches mentioned in this table.

Any reason to also not sort these?

Author: I let both of these slip through when fighting with the table format, I'll clear this up

Author: Fixed it, it should be ok here

.. _make_hgrid_yaml:

make_hgrid
==========

Structured YAML to run :ufs-utils:`make_hgrid<make-hgrid>` is validated by JSON Schema and requires the ``make_hgrid:`` block, described below. If ``make_hgrid`` is to be run via a batch system, the ``platform:`` block, described :ref:`here <platform_yaml>`, is also required.

> YAML keys and their descriptions are taken largely verbatim from the ``make_hgrid`` tool source code, available `here <https://github.com/NOAA-GFDL/FRE-NCtools/blob/65c34e76cfa2ff567f70da83f44cd6f991315f08/tools/make_hgrid/make_hgrid.c>`_.

Reviewer: Is that very-specific URL preferable to the simpler https://github.com/NOAA-GFDL/FRE-NCtools/blob/master/tools/make_hgrid/make_hgrid.c ? Maybe so, if it somehow ties the docs to the specific version of `make_hgrid` being used by... well, SRW? This leads into the fraught topic of versioning drivers, docs, etc. based on the versions of apps and components. Maybe it would be better to used the unversioned URL until we know more about version-pinning requirements?

Author: Good point, thats safer for sure. As long as we're just doing references and not pulling full text, this should be ok

Here is a prototype UW YAML ``make_hgrid:`` block, explained in detail below:

.. highlight:: yaml
.. literalinclude:: ../../../../shared/make_hgrid.yaml

UW YAML for the ``make_hgrid:`` Block
-------------------------------------

execution:
^^^^^^^^^^

See :ref:`here <execution_yaml>` for details.

run_dir:
^^^^^^^^

The path to the directory where ``make_hgrid`` will write its outputs.

config:
^^^^^^^

Describes the required parameters to run a ``make_hgrid`` configuration. 

grid_type:
""""""""""

> ``grid_type`` is the primary required key that defines the grid and additional required keys. It is specified with one of the following recognized key words:

Reviewer: It's so awful that I hesitate to mention it, but: Our schema currently ensures that, when `grid_type` is set to some value, that certain other `config` items are defined as well. But it does not, I think, ensure that _other_ values are unset -- and, if they are, I think they'll get translated into `--cli_flags` too, possibly resulting in invalid run commands. Maybe not: Maybe the `make_hgrid` program is resilient against this and ignores irrelevant flags. Hopefully! It might be possible to model this stricter use case with `"additionalProperties": false` lines in the `"then"` clauses we already have, or perhaps using a `oneOf` clause instead of `allOf`, to make sure that exactly one valid set of arguments is specified. It might be interesting to do a quick manual check to see how `make_hgrid` deal with an extra, inapplicable argument, so set our minds (or at least mine) at ease in case it behaves graceully. Otherwise, we could revisit this topic if/when a user runs into difficulty.

Author: To make this more unnerving, its also not especially clear which parameters are not "required" but are still necessary anyway... unfortunately, it may be up to the users to give us more examples

Author: I've parsed it to pseudocode and added that to my summary doc, working on converting that to even more conditionals

Author: Saving further details for future work

.. list-table:: Grid Types and Requirements
   :widths: 20 80
   :header-rows: 1

   * - Grid Type
     - Requirements

>    * - beta_plane_grid

Reviewer: ```suggestion
   * - ``beta_plane_grid``
```
Suggest formatting the entries in this column as code, too.

.. _make_hgrid_yaml:

make_hgrid
==========

Structured YAML to run :ufs-utils:`make_hgrid<make-hgrid>` is validated by JSON Schema and requires the ``make_hgrid:`` block, described below. If ``make_hgrid`` is to be run via a batch system, the ``platform:`` block, described :ref:`here <platform_yaml>`, is also required.

> .. important::

Reviewer 2: I like the green more!

.. _make_hgrid_yaml:

make_hgrid
==========

Structured YAML to run :ufs-utils:`make_hgrid<make-hgrid>` is validated by JSON Schema and requires the ``make_hgrid:`` block, described below. If ``make_hgrid`` is to be run via a batch system, the ``platform:`` block, described :ref:`here <platform_yaml>`, is also required.

> 

Reviewer 2: Please reference the source code where this documentation was taken from. We should probably be explicit that a lot of it is taken verbatim.

Author: Gotcha, text and link added

Here is a prototype UW YAML ``make_hgrid:`` block, explained in detail below:

.. highlight:: yaml
.. literalinclude:: ../../../../shared/make_hgrid.yaml

UW YAML for the ``make_hgrid:`` Block
-------------------------------------

execution:
^^^^^^^^^^

See :ref:`here <execution_yaml>` for details.

run_dir:
^^^^^^^^

The path to the directory where ``make_hgrid`` will write its outputs.

config:

> ^^^^^^^

Reviewer: Any reason not to have the items under `config:` and in the corresponding `make_hgrid.yaml` file sorted? Maybe there's a reason for the current order, but I couldn't guess it.

Author: I've listed them in order of the `make_hgrid` function docs, but I can resort them to something actually reasonable!

Author: I'm somewhat inclined to put grid_type first as it defines the subsequent parameters, and maintaining the similar parameter groupings like nx/ny - then most of the alphabetical order is already broken, but I can improve some of it

Reviewer: I think that'd be good if accompanied by a little additional text stating e.g. that `grid_type` is kind of a primary choice, and that the options that follow depend on that choice. In fact, is `grid_type` perhaps even required? Does it make sense not to have it?

Maybe we could present a table visually mapping `grid_type` values to the options that then become required for and are related to them.

Author: As reluctant as I am to update the jsonschema and its test again, I think that might be the way to go.

Author: AH, it is already required - I'm starting to lose track of all the conditions! I'll think about a visual map

Author: I've made sure all the key names have double-backstrokes and have used the table for `grid_type`, this helps a lot without adding further diagrams here

Describes the required parameters to run a ``make_hgrid`` configuration. 

grid_type:

> """"""""""

Reviewer: It's not a problem specific to this PR, but I think the visual difference between
```
config:
^^^^^^
```
and
```
grid_type:
"""""""""
```
is negligible to non-existent. Looking at the rendered HTML, it's not at all clear to me that `grid_type` et al. are subordinate to `config`. Possibly we should reconsider our formatting to make this clearer. Maybe we could discuss this in an upcoming DSM or Group.

> Specify type of topography. See the following options for more details:

Reviewer 2: ```suggestion
Specify grid type with one of the following recognized key words:
```

* from_file:              --my_grid_file must be specified.
* spectral_grid:          no other optional or required arguments.
* regular_lonlat_grid:    --nxbnds, --nybnds --xbnds, --ybnds, must be specified to define the grid bounds.
* tripolar_grid:          --nxbnds, --nybnds, --xbnds, --ybnds, must be specified to define the grid bounds.
* conformal_cubic_grid:   --nratio is optional argument.
* gnomonic_ed:            equal distance gnomonic cubic grid.
* simple_cartesian_grid:  --xbnds, --ybnds must be specified to define the grid bounds location and grid size. --simple_dx and --simple_dy must be specified to specify uniform cell length.

> * f_plane_grid:           For setting geometric fractors according f-plane. f_plane_latitude must be specified

Author: Thanks for catching the rest of the 'interesting' spelling left from the source!

> * f_plane_grid:           For setting geometric fractors according f-plane. f_plane_latitude must be specified

Reviewer 2: Also, should that be factor or fractor. Seems like maybe a typo (inherited from the code)?

> * f_plane_grid:           For setting geometric fractors according f-plane. f_plane_latitude need to be specified

Reviewer 3: ```suggestion
* f_plane_grid:           For setting geometric fractors according f-plane. f_plane_latitude must be specified
```

> * beta_plane_grid:        For setting geometric fractors according to beta plane. f_plane_latitude need to be specified 

Reviewer 2: The white space in this section is not preserved when rendered. You might consider a table. Please also use the double backtick highlighting for any of the flags. The use of colons here also makes me think these should be entries in the YAML instead of punctuation.

Author: Thanks - the table format looks much better!

xbnds:
""""""

Specify boundaries for defining zonal regions of varying resolution. When --tripolar is present, also defines the longitude of the two new poles.

ybnds:
""""""

Specify boundaries for defining meridional regions of varying resolution.

nlon:
"""""

Number of model grid points(supergrid) for each zonal regions of varying resolution.

nlat:
"""""

Number of model grid points(supergrid) for each meridinal regions of varying resolution.

dlon:
"""""

Nominal resolution of zonal regions.

dlat:
"""""

Nominal resolution of meridional regions.

lat_join:
"""""""""

Specify latitude for joining spherical and rotated bipolar grid. Default value is 65 degree

nratio:
"""""""

Specify the refinement ratio when calculating cell length and area of supergrid.

simple_dx:
""""""""""

Specify the uniform cell length in x-direction for simple cartesian grid.

simple_dy:
""""""""""

Specify the uniform cell length in y-direction for simple cartesian grid.

grid_name:
""""""""""

Specify the grid name.

center:
"""""""

Specify the center location of grid.

shift_fac:
""""""""""

Shift west by 180/shift_fac. Default value is 18.

do_schmidt:
"""""""""""

> Set to do Schmidt transformation to create astretched grid. When do_schmidt is set, the following must be set: --stretch_factor, --target_lon and --target_lat.

Reviewer 4: ```suggestion
Set to do Schmidt transformation to create a stretched grid. When do_schmidt is set, the following must be set: --stretch_factor, --target_lon and --target_lat.
```

nest_grids:
"""""""""""

Set to create this # nested grids as well as the global grid.

parent_tile:
""""""""""""

Specify the list of the parent tile number(s) of nest grid(s).

refine_ratio:
"""""""""""""

Specify the list of refinement ratio(s) for nest grid(s).

halo:
"""""

Halo size is to make sure the nest, including the halo region, is fully contained within a single parent (coarse) tile.

istart_nest:
""""""""""""

Specify the list of starting i-direction index(es) of nest grid(s) in parent tile supergrid(Fortran index).

iend_nest:
""""""""""

Specify the list of ending i-direction index(es) of nest grids in parent tile supergrid(Fortran index).

jstart_nest:
""""""""""""

Specify the list of starting j-direction index(es) of nest grids in parent tile supergrid(Fortran index). 

jend_nest:
""""""""""

Specify the list of ending j-direction index(es) of nest grids in parent tile supergrid(Fortran index).

great_circle_algorithm:
"""""""""""""""""""""""

When specified, great_circle_algorithm will be used to compute grid cell area. 

out_halo:
"""""""""

Extra halo size data to be written out.

angular_midpoint:
"""""""""""""""""

> If specified, when grid_type is from_file and the input is a NetCDF file, then the supergrid face midpoint coordinates are simply and independely (the lat independently from the lon) calculated as simple angular midpoints from the model grid coordiantes.

Reviewer 4: ```suggestion
If specified, when grid_type is from_file and the input is a NetCDF file, then the supergrid face midpoint coordinates are simply and independently (the lat independently from the lon) calculated as simple angular midpoints from the model grid coordinates.
```

non_length_angle:
"""""""""""""""""

When specified, will not output length(dx,dy) and angle (angle_dx, angle_dy).

rotate_poly:
""""""""""""

Set to calculate polar polygon areas by calculating the area of a copy of the polygon, with the copy  being rotated far away from the pole.

> verbose:

Reviewer 3: Is this an option that a user can add in the `config` block as a key? I didn't see it as an option in the example block provided, and wonder if it is does this add confusion to a user who will be using uwtools and see our verbose flag?  If it is an option for the `config` block should we add it in the YAML.

Author: Its a parameter for `make_hgrid`, so I put it as a key here, although its not required. Thats a valid concern, tho, I'm not exactly sure how to combine the two

Author: @maddenp-noaa I'm pondering the use case here, I'm not sure this flag is ever used by any of the apps. Would it be worth it to remove this as a YAML key and update the cli to take `--verbose` and driver itself to add `--verbose` to the runscript vars if used?

Reviewer: I think that verbose behavior or `uw` and verbose behavior of `make_hgrid` should be kept orthogonal so that users can choose which (or both, or neither) they want. `uw --verbose` controls the former, so I'd assume that a `verbose` UW YAML key for `make_hgrid` would control the latter. If `make_hgrid` takes a verbose flag, I think they YAML entry would be the correct way to control that.

""""""""

Will print out running time message when this is set. Otherwise the run will be silent when there is no error.

Example UW YAML With SRW YAML References

> ----------------------------------------

Reviewer 2: I would prefer we leave this section out. The point of a portion of our work is to refactor how SRW does this, in which case none of these environment variable will exist, and none of our tools support the syntax used here.

Author: Thats fair - I was torn between that, and wanting to help users with the remapping thats not necessarily clear without deeper references

Author: I've kept it as a note if needed later and removed it from here
